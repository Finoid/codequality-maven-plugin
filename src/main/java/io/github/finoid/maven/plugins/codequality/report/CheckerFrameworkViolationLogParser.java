package io.github.finoid.maven.plugins.codequality.report;

import org.codehaus.plexus.logging.Logger;
import org.codehaus.plexus.logging.console.ConsoleLogger;
import io.github.finoid.maven.plugins.codequality.exceptions.ParseException;
import io.github.finoid.maven.plugins.codequality.step.ViolationConverter;
import io.github.finoid.maven.plugins.codequality.util.Precondition;

import javax.inject.Inject;
import javax.inject.Singleton;
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStream;
import java.io.InputStreamReader;
import java.nio.charset.StandardCharsets;
import java.util.ArrayList;
import java.util.List;
import java.util.Optional;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

@Singleton
public class CheckerFrameworkViolationLogParser implements ViolationLogParser {
    public static final Pattern PART_VIOLATION_PATTERN = Pattern.compile("^(.*):(\\[\\d+,\\d+\\]).*(\\[.*\\])");
    public static final Pattern VIOLATION_PATTERN =
        Pattern.compile("^(?<path>.*):\\[(?<line>\\d+),(?<column>\\d+)\\].*\\[(?<rule>.*)\\] (?<description>.*)$");

    private static final Logger LOGGER = new ConsoleLogger(1, "console");

    private final ViolationConverter violationConverter;

    @Inject
    public CheckerFrameworkViolationLogParser(final ViolationConverter violationConverter) {
        this.violationConverter = Precondition.nonNull(violationConverter, "ViolationConverter shouldn't be null");
    }

    @Override
    public List<Violation> parse(final InputStream fileInputStream) {
        final List<Violation> violations = new ArrayList<>();

        try (final BufferedReader reader = new BufferedReader(new InputStreamReader(fileInputStream, StandardCharsets.UTF_8))) {
            String nullableLine;

            while ((nullableLine = reader.readLine()) != null) {
                if (isNotCheckerFrameworkViolationLine(nullableLine)) {
                    continue;
                }

                parseViolation(nullableLine)
                    .ifPresent(violations::add);
            }

            return violations;
        } catch (final IOException e) {
            throw new ParseException("Exception during parsing", e);
        }
    }

    private Optional<Violation> parseViolation(final String currentLine) {
        final Matcher violationMatcher = VIOLATION_PATTERN.matcher(currentLine);
        if (!violationMatcher.find()) {
            LOGGER.warn("Unexpected checker framework log. Log: " + currentLine);

            return Optional.empty();
        }

        return violationOf(currentLine);
    }

    private Optional<Violation> violationOf(final String checkerFrameworkViolationLines) {
        final Matcher violationMatcher = VIOLATION_PATTERN.matcher(checkerFrameworkViolationLines);

        if (!violationMatcher.find()) {
            LOGGER.warn("Unexpected checker framework log. Log: " + checkerFrameworkViolationLines);

            return Optional.empty();
        }

        return Optional.of(violationConverter.ofCheckerFrameworkViolationMatcher(violationMatcher));
    }

    private static boolean isNotCheckerFrameworkViolationLine(final String nullableLine) {
        final Matcher partViolationMatcher = PART_VIOLATION_PATTERN.matcher(nullableLine);

        return !partViolationMatcher.find();
    }

}
