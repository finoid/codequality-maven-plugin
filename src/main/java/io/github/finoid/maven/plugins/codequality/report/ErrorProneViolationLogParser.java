package io.github.finoid.maven.plugins.codequality.report;

import io.github.finoid.maven.plugins.codequality.exceptions.ParseException;
import io.github.finoid.maven.plugins.codequality.step.ViolationConverter;
import io.github.finoid.maven.plugins.codequality.util.Precondition;
import org.codehaus.plexus.logging.Logger;
import org.codehaus.plexus.logging.console.ConsoleLogger;

import javax.inject.Inject;
import javax.inject.Singleton;
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStream;
import java.io.InputStreamReader;
import java.nio.charset.StandardCharsets;
import java.util.ArrayList;
import java.util.List;
import java.util.Optional;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

@Singleton
public class ErrorProneViolationLogParser implements ViolationLogParser {
    public static final Pattern VIOLATION_PATTERN =
        Pattern.compile("^(?<path>.*):\\[(?<line>\\d+)(?:,(?<column>\\d+))?\\] \\[(?<rule>.*)\\] (?<description>.*\\s*\\(.*\\s*.*)$");

    private static final Pattern PART_VIOLATION_PATTERN = Pattern.compile("^(.*):\\[(\\d+(?:,\\d+)?)\\] \\[(.*?)\\]");
    private static final Logger LOGGER = new ConsoleLogger(1, "console");
    private static final String THIRD_LINE_PREFIX = "  Did you mean";
    private static final int READ_AHEAD_LIMIT = 8 * 1024;

    private final ViolationConverter violationConverter;

    @Inject
    public ErrorProneViolationLogParser(final ViolationConverter violationConverter) {
        this.violationConverter = Precondition.nonNull(violationConverter, "ViolationConverter shouldn't be null");
    }

    @Override
    public List<Violation> parse(final InputStream fileInputStream) {
        final List<Violation> violations = new ArrayList<>();

        try (final BufferedReader reader = new BufferedReader(new InputStreamReader(fileInputStream, StandardCharsets.UTF_8))) {
            String nullableLine;

            while ((nullableLine = reader.readLine()) != null) {
                if (isNotErrorProneViolationLine(nullableLine)) {
                    continue;
                }

                parseViolation(nullableLine, reader)
                    .ifPresent(violations::add);
            }

            return violations;
        } catch (final IOException e) {
            throw new ParseException("Exception during parsing", e);
        }
    }

    private Optional<Violation> parseViolation(final String currentLine, final BufferedReader reader) throws IOException {
        final StringBuilder errorProneViolationBuffer = new StringBuilder();

        errorProneViolationBuffer
            .append(currentLine)
            .append(reader.readLine());

        reader.mark(READ_AHEAD_LIMIT);

        String peakAhead = reader.readLine();

        // optional third line: typically "  Did you mean ..."
        if (peakAhead.startsWith(THIRD_LINE_PREFIX)) {
            errorProneViolationBuffer.append(peakAhead);
        } else {
            // Not a valid third line, rewind so the outer loop will process it
            reader.reset();
        }

        final Matcher violationMatcher = VIOLATION_PATTERN.matcher(errorProneViolationBuffer.toString());
        if (!violationMatcher.find()) {
            LOGGER.warn("Unexpected error prone log. Log: " + errorProneViolationBuffer);

            return Optional.empty();
        }

        return violationOf(errorProneViolationBuffer.toString());
    }

    private Optional<Violation> violationOf(final String errorProneViolationLines) {
        final Matcher violationMatcher = VIOLATION_PATTERN.matcher(errorProneViolationLines);

        if (!violationMatcher.find()) {
            LOGGER.warn("Unexpected error prone log. Log: " + errorProneViolationLines);

            return Optional.empty();
        }

        return Optional.of(violationConverter.ofErrorProneViolationMatcher(violationMatcher));
    }

    private static boolean isNotErrorProneViolationLine(final String nullableLine) {
        final Matcher partViolationMatcher = PART_VIOLATION_PATTERN.matcher(nullableLine);

        return !partViolationMatcher.find();
    }
}
