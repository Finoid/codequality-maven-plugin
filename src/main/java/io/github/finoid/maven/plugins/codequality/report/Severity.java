package io.github.finoid.maven.plugins.codequality.report;

public enum Severity {
    INFO,
    MINOR,
    MAJOR,
    CRITICAL,
    BLOCKER;

    @SuppressWarnings("EnumOrdinal")
    public boolean isHigherThanOrEqual(final Severity severity) {
        return this.ordinal() >= severity.ordinal();
    }
}
