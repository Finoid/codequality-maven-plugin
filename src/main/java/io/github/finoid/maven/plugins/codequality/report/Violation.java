package io.github.finoid.maven.plugins.codequality.report;

import lombok.Value;

@Value
public class Violation {
    String description;
    String fingerprint;
    Severity severity;
    String relativePath;
    String fullPath;
    Integer line;
    Integer columnNumber;
    String rule;
}
