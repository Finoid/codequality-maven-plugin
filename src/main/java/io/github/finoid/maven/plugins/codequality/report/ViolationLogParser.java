package io.github.finoid.maven.plugins.codequality.report;

import io.github.finoid.maven.plugins.codequality.exceptions.ParseException;

import java.io.InputStream;
import java.util.List;

public interface ViolationLogParser {

    /**
     * Parses the provided input stream to extract a list of violations from log entries.
     *
     * @param fileInputStream the input stream containing the log data to be parsed.
     * @return a list of {@link Violation} objects representing the identified violations in the logs.
     * @throws ParseException if an error occurs during the parsing process, such as invalid log formatting or issues reading from the input stream.
     */
    List<Violation> parse(final InputStream fileInputStream);

}
