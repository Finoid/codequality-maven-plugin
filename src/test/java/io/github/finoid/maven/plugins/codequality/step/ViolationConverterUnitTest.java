package io.github.finoid.maven.plugins.codequality.step;

import io.github.finoid.maven.plugins.codequality.fixtures.AuditEventFaker;
import io.github.finoid.maven.plugins.codequality.fixtures.UnitTest;
import io.github.finoid.maven.plugins.codequality.report.CheckerFrameworkViolationLogParser;
import io.github.finoid.maven.plugins.codequality.report.ErrorProneViolationLogParser;
import org.apache.maven.project.MavenProject;
import org.junit.jupiter.api.Test;
import org.mockito.InjectMocks;
import org.mockito.Mock;
import org.mockito.Mockito;

import java.nio.file.Path;
import java.nio.file.Paths;
import java.util.regex.Matcher;

class ViolationConverterUnitTest extends UnitTest {
    private static Path WORKING_DIRECTORY = Paths.get("")
        .toAbsolutePath();

    @Mock
    private MavenProject project;
    @InjectMocks
    private ViolationConverter unit;

    @Test
    void givenValidLogEntryAndMatcher_whenOfAuditEvent_thenExpectedViolation() {
        Mockito.when(project.getBasedir())
            .thenReturn(WORKING_DIRECTORY.toFile());

        var event = AuditEventFaker.auditEvent()
            .withFileName(WORKING_DIRECTORY + "/src/main/java/io/github/finoid/maven/plugins/codequality"
                          + "/step/CheckerFrameworkStep.java")
            .create();

        var violation = unit.ofAuditEvent(event);

        snapshot(violation, "*..fullPath");
    }

    @Test
    void givenValidLogEntryAndMatcher_whenOfCheckerFrameworkViolationMatcher_thenExpectedViolation() {
        final Matcher violationMatcher = CheckerFrameworkViolationLogParser.VIOLATION_PATTERN.matcher(
            WORKING_DIRECTORY
            + "/src/main/java/io/github/finoid/maven/plugins/codequality/step/CheckerFrameworkStep.java:[10,29] error: [required.method.not.called] "
            + "@MustCall method close may not have been invoked on SpringApplication.run(Application.class, args) or any of its aliases.");
        Mockito.when(project.getBasedir())
            .thenReturn(WORKING_DIRECTORY.toFile());

        violationMatcher.find();

        var violation = unit.ofCheckerFrameworkViolationMatcher(violationMatcher);

        snapshot(violation, "*..fullPath");
    }

    @Test
    void givenValidLogEntryAndMatcher_whenOfErrorProneViolationMatcher_thenExpectedViolation() {
        final Matcher violationMatcher = ErrorProneViolationLogParser.VIOLATION_PATTERN.matcher(
            WORKING_DIRECTORY
            + "/src/main/java/io/github/finoid/library/otel/OpenTelemetryUtils.java:[24,24] [StringCaseLocaleUsage] Specify a `Locale` when calling "
            + "`String#to{Lower,Upper}Case`. (Note: there are multiple suggested fixes; the third may be most appropriate if you're dealing with "
            + "ASCII Strings.)\n"
            + "    (see https://errorprone.info/bugpattern/StringCaseLocaleUsage)\n"
            + "  Did you mean '.toLowerCase(Locale.ROOT);' or '.toLowerCase(Locale.getDefault());' or "
            + "'return Ascii.toLowerCase(currentSpan.getSpanContext()'?");
        Mockito.when(project.getBasedir())
            .thenReturn(WORKING_DIRECTORY.toFile());

        violationMatcher.find();

        var violation = unit.ofErrorProneViolationMatcher(violationMatcher);

        snapshot(violation, "*..fullPath");
    }
}