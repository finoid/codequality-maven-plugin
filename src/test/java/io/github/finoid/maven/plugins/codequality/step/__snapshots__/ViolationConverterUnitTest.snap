io.github.finoid.maven.plugins.codequality.step.ViolationConverterUnitTest.givenValidLogEntryAndMatcher_whenOfAuditEvent_thenExpectedViolation=[
  {
    "columnNumber" : 14,
    "description" : "Checkstyle: Descriptive violation siTgN",
    "fingerprint" : "2ada6236b7b78f6b4b14e0f6eedd625e489bfb37af8525e055165f79bf6cd1b7",
    "fullPath" : "***MASKED***",
    "line" : 54,
    "relativePath" : "src/main/java/io/github/finoid/maven/plugins/codequality/step/CheckerFrameworkStep.java",
    "rule" : "LX8lv - Violation",
    "severity" : "INFO"
  }
]


io.github.finoid.maven.plugins.codequality.step.ViolationConverterUnitTest.givenValidLogEntryAndMatcher_whenOfCheckerFrameworkViolationMatcher_thenExpectedViolation=[
  {
    "columnNumber" : 29,
    "description" : "CheckerFramework: @MustCall method close may not have been invoked on SpringApplication.run(Application.class, args) or any of its aliases.",
    "fingerprint" : "a74302f3e5535ae3ccd8af1ba7ae87c16f4d27faba4f3b6539106de5ac3f2da1",
    "fullPath" : "***MASKED***",
    "line" : 10,
    "relativePath" : "src/main/java/io/github/finoid/maven/plugins/codequality/step/CheckerFrameworkStep.java",
    "rule" : "required.method.not.called",
    "severity" : "MINOR"
  }
]


io.github.finoid.maven.plugins.codequality.step.ViolationConverterUnitTest.givenValidLogEntryAndMatcher_whenOfErrorProneViolationMatcher_thenExpectedViolation=[
  {
    "columnNumber" : 24,
    "description" : "ErrorProne: Specify a `Locale` when calling `String#to{Lower,Upper}Case`. (Note: there are multiple suggested fixes; the third may be most appropriate if you're dealing with ASCII Strings.)\n    (see https://errorprone.info/bugpattern/StringCaseLocaleUsage)\n  Did you mean '.toLowerCase(Locale.ROOT);' or '.toLowerCase(Locale.getDefault());' or 'return Ascii.toLowerCase(currentSpan.getSpanContext()'?",
    "fingerprint" : "16251fd6d9cde2f70cb271982ecc4ef4afd91001300ff80116e66ec8ad65b90b",
    "fullPath" : "***MASKED***",
    "line" : 24,
    "relativePath" : "src/main/java/io/github/finoid/library/otel/OpenTelemetryUtils.java",
    "rule" : "StringCaseLocaleUsage",
    "severity" : "MINOR"
  }
]